samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreRestoreManager.java

```diff
@@ -0,0 +1,349 @@
+    }).join(); // TODO BLOCKER dchen1 make non-blocking.
+  }
+
+  /**
+   * Determines if the store needs to be restored from remote snapshot based on local and remote state.
+   */
+  @VisibleForTesting
+  static boolean shouldRestore(String taskName, String storeName, DirIndex dirIndex,
+      Path storeCheckpointDir, StorageConfig storageConfig, BlobStoreUtil blobStoreUtil) {
+    // if a store checkpoint directory exists for the last successful task checkpoint, try to use it.
+    boolean restoreStore;
+    if (Files.exists(storeCheckpointDir)) {
+      if (storageConfig.getCleanLoggedStoreDirsOnStart(storeName)) {
+        LOG.debug("Restoring task: {} store: {} from remote snapshot since the store is configured to be " +
+            "restored on each restart.", taskName, storeName);
+        restoreStore = true;
+      } else if (blobStoreUtil.areSameDir(FILES_TO_IGNORE, true).test(storeCheckpointDir.toFile(), dirIndex)) {
+        restoreStore = false; // no restore required for this store.
+      } else {
+        // we don't optimize for the case when the local host doesn't contain the most recent store checkpoint
+        // directory but contains an older checkpoint directory which could have partial overlap with the remote
+        // snapshot. we also don't try to optimize for any edge cases where the most recent checkpoint directory
+        // contents could be partially different than the remote store (afaik, there is no known valid scenario
+        // where this could happen right now, except for the offset file handling above).
+        // it's simpler and fast enough for now to restore the entire store instead.
+
+        LOG.error("Local store checkpoint directory: {} contents are not the same as the remote snapshot. " +
+            "Queuing for restore from remote snapshot.", storeCheckpointDir);
+        // old checkpoint directory will be deleted later during commits
+        restoreStore = true;
+      }
+    } else { // did not find last checkpoint dir, restore the store from the remote blob store
+      LOG.debug("No local store checkpoint directory found at: {}. " +
+          "Queuing for restore from remote snapshot.", storeCheckpointDir);
+      restoreStore = true;
+    }
+
+    return restoreStore;
+  }
+
+  /**
+   * Starts the restore for the store, enqueuing all restore-completion futures into {@param restoreFutures}.
+   */
+  @VisibleForTesting
+  static void enqueueRestore(String jobName, String jobId, String taskName, String storeName, File storeDir, DirIndex dirIndex,
+      long storeRestoreStartTime, List<CompletionStage<Void>> restoreFutures,
+      BlobStoreUtil blobStoreUtil, BlobStoreRestoreManagerMetrics metrics, ExecutorService executor) {
+    metrics.storePreRestoreNs.get(storeName).set(System.nanoTime() - storeRestoreStartTime);
+
+    Metadata requestMetadata = new Metadata(storeDir.getAbsolutePath(), Optional.empty(), jobName, jobId, taskName, storeName);
```

Comment:
Metadata contains Metadata related to the current request. So a restoreDir request has storeDir in payload. When the restoreDir is called, it creates new Metadata object with payload as individual file and that's what is eventually sent to the blob store API (GET). Basically the real request to blob store has file names, because at every step right request Metadata object is built. That way, we track request at each step and use that to log if necessary.